core/src/main/scala/kafka/raft/KafkaMetadataLog.scala

```diff
@@ -48,6 +48,17 @@ final class KafkaMetadataLog private (
 
   this.logIdent = s"[MetadataLog partition=$topicPartition, nodeId=${config.nodeId}] "
 
+  // If the log start offset is not 0, then we must have a snapshot which covers the
+  // initial state up to the current log start offset.
+  if (log.logStartOffset > 0) {
+    val latestSnapshotId = snapshots.lastOption.map(_._1)
+    if (!latestSnapshotId.exists(snapshotId => snapshotId.offset >= log.logStartOffset)) {
+      throw new IllegalStateException("Inconsistent snapshot state: there must be a snapshot " +
+        s"at an offset larger then the current log start offset ${log.logStartOffset}, but the " +
+        s"latest snapshot is ${latestSnapshotId}")
+    }
+  }
+
```

Comment:
What do you think about moving this check to `recoverSnapshots` or before calling `recoverSnapshots`?

`recoverSnapshots` makes changes to the log dir. Kafka should probably validate the log dir before making changes to it?

Reply:
Makes sense.

Reply:
I guess a specific case where doing this prior to `recoverSnapshots` might help is if we have a deleted file corresponding to a snapshot in the range we are looking for.

Reply:
Yes. `recoverSnapshots` doesn't make a log dir invalid but it may make it "more" invalid by deleting more snapshots.